lib/src/main/java/org/kpipe/consumer/FunctionalConsumer.java

```diff
@@ -648,7 +745,13 @@ protected void processRecords(final ConsumerRecords<K, V> records) {
         .stream(records.records(topic).spliterator(), false)
         .forEach(record -> {
           try {
-            virtualThreadExecutor.submit(() -> processRecord(record));
+            virtualThreadExecutor.submit(() -> {
+              // Track offset before processing
+              if (offsetManager != null) {
+                commandQueue.offer(ConsumerCommand.TRACK_OFFSET.withRecord(record));
+              }
+              processRecord(record);
+            });
```

Comment:
the offset tracking should be done before submission into virtual threadExecutor - its not intensive (just adding to queue) but ensures that tracking is done sequentially - so no need to worry about lower offsets being added after higher offsets to pending queue. 
That simplifies reasoning and handling of commits as naturally the 'polled' offsets can be thought off 'pending' regardless of whether virtual thread started processing on them or not...

Reply:
nice catch 🙏 

Fixed